examples/char-rnn.cpp

```diff
@@ -0,0 +1,260 @@
+llvm::cl::opt<unsigned>
+    generateChars("chars", llvm::cl::desc("Generate this number of chars."),
+                  llvm::cl::init(1000), llvm::cl::value_desc("N"),
+                  llvm::cl::cat(category));
+
+} // namespace
+
+PseudoRNG RNG;
+
+/// Clip the value \p c to the range 0..127, which is standard ascii.
+static size_t clipASCII(char c) {
+  size_t c1 = c;
+  if (c1 > 127) {
+    c1 = 127;
+  }
+  return c1;
+}
+
+/// Load text into \p inputText that has the format  [B, S, 128], where B is
+/// the batch size, S is the length of the sentence, and 128 is the one-hot
+/// representation of the text. Load the expected index into \p nextChar that
+/// has the format [B, S], where each element is the softmax index of the
+/// next char. If \p train is false then only load the first slice of inputText.
+static void loadText(Tensor &inputText, Tensor &nextChar, llvm::StringRef text,
+                     bool train) {
+  assert(text.size() > 2 && "The buffer must contain at least two chars");
+  inputText.zero();
+  nextChar.zero();
+
+  auto idim = inputText.dims();
+  assert(idim.size() == 3 && "invalid input tensor");
+  auto B = idim[0];
+  auto S = idim[1];
+
+  auto IH = inputText.getHandle();
+  auto NH = nextChar.getHandle<size_t>();
+
+  // Fill the tensor with slices from the sentence with an offset of 1.
+  // Example:
+  //  |Hell|o| World
+  //  |ello| |World
+  //  |llo |W|orld
+  //  |lo W|o|rld
+  for (size_t i = 0; i < B; i++) {
+    for (size_t j = 0; j < S; j++) {
+      size_t c = clipASCII(text[i + j]);
+
+      IH.at({i, j, c}) = 1.0;
+      if (train) {
+        size_t c1 = clipASCII(text[i + j + 1]);
+        NH.at({i, j}) = c1;
+      }
+    }
+
+    // Only load the first slice in the batch when in inference mode.
+    if (!train) {
+      return;
+    }
+  }
+}
+
+/// This method selects a random number based on a softmax distribution. One
+/// property of this distribution is that the sum of all probabilities is equal
+/// to one. The algorithm that we use here picks a random number between zero
+/// and one. Then, we scan the tensor and accumulate the probabilities. We stop
+/// and pick the index when sum is greater than the selected random number.
+static char getPredictedChar(Tensor &inputText, size_t slice, size_t word) {
+  auto IH = inputText.getHandle();
+
+  // Pick a random number between zero and one.
+  double x = std::abs(RNG.nextRand());
+  double sum = 0;
+  // Accumulate the probabilities into 'sum'.
+  for (size_t i = 0; i < 128; i++) {
+    sum += IH.at({slice, word, i});
+    // As soon as we cross the threashold return the index.
+    if (sum > x) {
+      return i;
+    }
+  }
+  return 127;
+}
+
+/// Loads the content of a file or stdin to a memory buffer.
+static std::unique_ptr<llvm::MemoryBuffer> loadFile(llvm::StringRef filename) {
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> fileBufOrErr =
+      llvm::MemoryBuffer::getFileOrSTDIN(filename);
+  if (!fileBufOrErr) {
```

Comment:
does it have any error code when fails?

Reply:
Yes. I added code to print the error. Example:

Error! Failed to open file: asdsd.txt
No such file or directory